core/src/core/pmmr/segment.rs

```diff
@@ -0,0 +1,519 @@
+			// Full segment: last position of segment is subtree root
+			hashes
+				.pop()
+				.flatten()
+				.ok_or_else(|| SegmentError::MissingHash(segment_last_pos))
+		} else {
+			// Not full (only final segment): peaks in segment, bag them together
+			let peaks = pmmr::peaks(last_pos)
+				.into_iter()
+				.filter(|&pos| pos >= segment_first_pos && pos <= segment_last_pos)
+				.rev();
+			let mut hash = None;
+			for pos in peaks {
+				let mut lhash = hashes.pop().ok_or_else(|| SegmentError::MissingHash(pos))?;
+				if lhash.is_none() && bitmap.is_some() {
+					// If this entire peak is pruned, load it from the segment hashes
+					lhash = Some(self.get_hash(pos)?);
+				}
+				let lhash = lhash.ok_or_else(|| SegmentError::MissingHash(pos))?;
+
+				hash = match hash {
+					None => Some(lhash),
+					Some(rhash) => Some((lhash, rhash).hash_with_index(last_pos)),
+				};
+			}
+			hash.ok_or_else(|| SegmentError::MissingHash(0))
+		}
+	}
+
+	/// Check validity of the segment by calculating its root and validating the merkle proof
+	pub fn validate(
+		&self,
+		last_pos: u64,
+		bitmap: Option<&Bitmap>,
+		mmr_root: Hash,
+	) -> Result<(), SegmentError> {
+		let (first, last) = self.segment_pos_range(last_pos);
+		let segment_root = self.root(last_pos, bitmap)?;
+		self.proof
+			.validate(last_pos, mmr_root, first, last, segment_root)
+	}
+}
+
+impl<T: Readable> Readable for Segment<T> {
+	fn read<R: Reader>(reader: &mut R) -> Result<Self, Error> {
+		let identifier = Readable::read(reader)?;
+
+		let mut last_pos = 0;
+
+		let n_hashes = reader.read_u64()? as usize;
+		let mut hashes = Vec::with_capacity(n_hashes);
+		for _ in 0..n_hashes {
+			let pos = reader.read_u64()?;
+			if pos <= last_pos {
+				return Err(Error::SortError);
+			}
+			last_pos = pos;
+			let hash: Hash = Readable::read(reader)?;
+			hashes.push((pos, hash));
+		}
+
+		let n_leaves = reader.read_u64()? as usize;
+		let mut leaf_data = Vec::with_capacity(n_leaves);
+		last_pos = 0;
+		for _ in 0..n_leaves {
+			let pos = reader.read_u64()?;
+			if pos <= last_pos {
+				return Err(Error::SortError);
+			}
+			last_pos = pos;
+			let data: T = Readable::read(reader)?;
+			leaf_data.push((pos, data));
+		}
+
+		let proof = Readable::read(reader)?;
+
+		Ok(Self {
+			identifier,
+			hashes,
+			leaf_data,
+			proof,
+		})
+	}
+}
+
+impl<T: Writeable> Writeable for Segment<T> {
+	fn write<W: Writer>(&self, writer: &mut W) -> Result<(), Error> {
+		Writeable::write(&self.identifier, writer)?;
+		writer.write_u64(self.hashes.len() as u64)?;
+		for (pos, hash) in &self.hashes {
+			writer.write_u64(*pos)?;
+			Writeable::write(hash, writer)?;
+		}
+		writer.write_u64(self.leaf_data.len() as u64)?;
+		for (pos, data) in &self.leaf_data {
+			writer.write_u64(*pos)?;
+			Writeable::write(data, writer)?;
+		}
+		Writeable::write(&self.proof, writer)?;
+		Ok(())
+	}
+}
+
+/// Merkle proof of a segment
+#[derive(Clone, Debug, Eq, PartialEq)]
+pub struct SegmentProof {
```

Comment:
How much overlap do we have with the existing `MerkleProof` struct here?
Have you considered potentially combining/merging/reusing this?

Reply:
There is some overlap but we have a few extra assumptions in here that change the calculation somewhat. I could try to see if we can somehow extract some common code without cluttering it.

Reply:
I could use the `MerkleProof` struct instead, but we can't really use any of the functions already defined in it. We would have to add new functions to generate/validate the proof for the segment. So I think it would be more clean to keep them separate for now.

However if we possibly want to merge them in the future, we should maybe add the `mmr_size` field to `SegmentProof`, to keep it compatible with `MerkleProof`. Right now it is supplied as an argument to the validation function and assumed to be obtained from the block header.